### PR TITLE
[feature][broker] add config maxUnloadBundleNumPerShedding for UniformLoadShedder

### PR DESCRIPTION
**Pull request by lordcheng10**
### Motivation
In order to control the number of unloaded bundles more accurately each time, the configuration maxUnloadBundleNumPerShedding is added.

The purpose of adding this configuration is to control the number of bundles unloaded at one time, and prevent too many bundles unloaded at one time, resulting in cluster instability.

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

**Comment by lordcheng10** [comment]
@rdhabalia PTAL,thanks!

**Comment by lordcheng10** [comment]
@rdhabalia PTAL,thanks!

**Comment by lordcheng10** [comment]
@eolivelli @hangc0276 @codelipenghui PTAL,thanks!

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java [hunk 1]
```diff
@@ -2085,6 +2085,13 @@ public class ServiceConfiguration implements PulsarConfiguration {
     )
     private double loadBalancerMsgThroughputMultiplierDifferenceShedderThreshold = 4;
 
+    @FieldContext(
+            dynamic = true,
+            category = CATEGORY_LOAD_BALANCER,
+            doc = "For each uniform balanced unload, the maximum number of bundles that can be unloaded."
+    )
+    private int maxUnloadBundleNumPerShedding = 2;
```
Comment (Demogorgon314) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:2093]: We may need to write a PIP to explain this change.

Comment (lordcheng10) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:2093]: The purpose of adding this configuration is to control the number of bundles unloaded at one time, and prevent too many bundles unloaded at one time, resulting in cluster instability.
@Demogorgon314

Comment (lordcheng10) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:2093]: This is just to add a configuration, the purpose is to control the maximum number of bundles in one round of equalization adjustment.

Comment (lordcheng10) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java]: @Demogorgon314 I changed the default value of this configuration to -1. By default, there is no limit to the maximum number of bundles that can be adjusted each time, so it has no effect on the original logic:
image

Comment (Demogorgon314) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java]: > The purpose of adding this configuration is to control the number of bundles unloaded at one time, and prevent too many bundles unloaded at one time, resulting in cluster instability.

Ok,  you can add this purpose to the `Motivation ` part. Also, I believe we still need to add the unit test.

Comment (lordcheng10) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java]: OK. I will fix

Comment (lordcheng10) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java]: fixed. PTAL,thanks! @Demogorgon314

**Comment by lordcheng10** [comment]
@codelipenghui @eolivelli @codelipenghui PTAL,thanks!

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/UniformLoadShedder.java [hunk 2]
```diff
@@ -144,6 +145,10 @@ public Multimap<String, String> findBundlesForUnloading(final LoadData loadData,
                             return Triple.of(bundle, bundleData, throughput);
                         }).filter(e -> !recentlyUnloadedBundles.containsKey(e.getLeft()))
                         .sorted((e1, e2) -> Double.compare(e2.getRight(), e1.getRight())).forEach((e) -> {
+                            if (conf.getMaxUnloadBundleNumPerShedding() != -1
+                                    && unloadBundleCount.getValue() >= conf.getMaxUnloadBundleNumPerShedding()) {
```
Comment (Demogorgon314) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/UniformLoadShedder.java:149]:
```suggestion
                                    && selectedBundlesCache.size()  >= conf.getMaxUnloadBundleNumPerShedding()) {
```
We don't need a new variable here.

Comment (lordcheng10) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/UniformLoadShedder.java]: fixed. PTAL,thanks!  @Demogorgon314

**Review by Demogorgon314 — APPROVED** [review]

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java [hunk 3]
```diff
@@ -2153,6 +2153,14 @@ public class ServiceConfiguration implements PulsarConfiguration {
     )
     private double loadBalancerMsgThroughputMultiplierDifferenceShedderThreshold = 4;
 
+    @FieldContext(
+            dynamic = true,
+            category = CATEGORY_LOAD_BALANCER,
+            doc = "For each uniform balanced unload, the maximum number of bundles that can be unloaded."
+                    + " The default value is -1, which means no limit"
+    )
+    private int maxUnloadBundleNumPerShedding = -1;
```
Comment (aloyszhang) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:2162]: It seems maxUnloadBundleNumPerShedding=0 also means no limit, right?

Comment (lordcheng10) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:2162]: 0 means unload cannot be performed. @aloyszhang

Comment (aloyszhang) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:2162]: Got it

**Review by aloyszhang — APPROVED** [review]
LGTM

**Comment by lordcheng10** [comment]
@codelipenghui @Technoboy- PTAL,thanks!

**Review by Technoboy- — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**